pkg/kv/kvserver/store_send.go

```diff
@@ -157,104 +162,145 @@ func (s *Store) Send(
-		// that the client requested, and all the ranges in between.
-		ri, err := t.MismatchedRange()
+	// Tracks suggested ranges to return to the caller. Suggested ranges are aggregated from
+	// two sources.
+	// (1): On a RangeKeyMismatchError that is retriable.
+	// (2): On a successful batch request, where suggested ranges are returned
+	//      by the replica to update the client with. This is appended before returning.
+	var candidateRanges []roachpb.RangeInfo
+
+	// Run a retry loop on retriable RangeKeyMismatchErrors, where the requested RangeID is
+	// no longer correct, however we have the Range locally and can redirect the request.
+	// TODO(kvoli): The number of retries is limited to at most two repeated mismatch ranges. Recurrent
+	// RangeKeyMismatchErrors shouldn't be possible unless repeated splitting upon the range containing
+	// the request key span occurs. This seems unlikely, however as this behavior is degenerate retries
+	// is limited to at most 3 before returning an error to the client. Establish an appropriate bounds.
+	for retries, isRetriableRangeKeyMismatch := 0, true; retries < maxRKeyMismatchRetries && isRetriableRangeKeyMismatch; retries++ {
+		// Get range and add command to the range for execution.
+		repl, err := s.GetReplica(ba.RangeID)
 		if err != nil {
 			return nil, roachpb.NewError(err)
 		}
-		skipRID := ri.Desc.RangeID // We already have info on one range, so don't add it again below.
-		startKey := ri.Desc.StartKey
-		if rSpan.Key.Less(startKey) {
-			startKey = rSpan.Key
+		if !repl.IsInitialized() {
+			repl.mu.RLock()
+			replicaID := repl.mu.replicaID
+			repl.mu.RUnlock()
+
+			// If we have an uninitialized copy of the range, then we are
+			// probably a valid member of the range, we're just in the
+			// process of getting our snapshot. If we returned
+			// RangeNotFoundError, the client would invalidate its cache,
+			// but we can be smarter: the replica that caused our
+			// uninitialized replica to be created is most likely the
+			// leader.
+			return nil, roachpb.NewError(&roachpb.NotLeaseHolderError{
+				RangeID:     ba.RangeID,
+				LeaseHolder: repl.creatingReplica,
+				// The replica doesn't have a range descriptor yet, so we have to build
+				// a ReplicaDescriptor manually.
+				Replica: roachpb.ReplicaDescriptor{
+					NodeID:    repl.store.nodeDesc.NodeID,
+					StoreID:   repl.store.StoreID(),
+					ReplicaID: replicaID,
+				},
+			})
 		}
-		endKey := ri.Desc.EndKey
-		if endKey.Less(rSpan.EndKey) {
-			endKey = rSpan.EndKey
+
+		isRetriableRangeKeyMismatch = false
+		br, pErr = repl.Send(ctx, ba)
+		if pErr == nil {
+			// After succeeding on a retry it is necessary to return the RangeInfos back
+			// to the client, to invalidate their cache. When the batch request has a
+			// mismatch and is not retriable, this information is instead returned
+			// inside pErr.RangeInfos.
+			if retries > 0 {
```

Comment:
Do we need this conditional? If there are not retries, I think our set of candidate range infos would be empty. This relates to the point above -- we could simplify this a bit if all the retry block did was accumulate the set of range infos gathered across all the retries, and only at the very end populate the right proto (BatchResponse or PErr).

Reply:
It is redundant. Updated to accumulate - the only thing that is a bit iffy here, for the approach of populating pErr at the end is the type of the error needs to be known. So doing this at the very end wouldn't make sense as it would require another switch. I've still updated it to be more explicit, however not all at the end.